File: leetcode/103_Binary_Tree_Zigzag_Level_Order_Traversal.py
from typing import List, Optional
import unittest

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __repr__(self) -> str:
        return str(self.val)

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        current_stack, next_stack = [root], []
        level = 0
        res = []
        while current_stack:
            node = current_stack.pop()
            if node is not None:
                while level >= len(res):
                    res.append([])
                res[level].append(node.val)
                if level % 2 == 0:
                    next_stack.append(node.left)
                    next_stack.append(node.right)
                else:
                    next_stack.append(node.right)
                    next_stack.append(node.left)
            if not current_stack:
                current_stack, next_stack = next_stack, current_stack
                level += 1
        return res

class ZigzagLevelOrderTest(unittest.TestCase):
    def test_1(self) -> None:
        self.assertEqual(Solution().zigzagLevelOrder(TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))), [[3],[20,9],[15,7]])
    
    def test_2(self) -> None:
        self.assertEqual(Solution().zigzagLevelOrder(TreeNode(1)), [[1]])

    def test_3(self) -> None:
        self.assertEqual(Solution().zigzagLevelOrder(None), [])
    

File: leetcode/103_Binary_Tree_Zigzag_Level_Order_Traversal.rs
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}
 
impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None
        }
  }
}

use std::rc::Rc;
use std::cell::RefCell;

struct Solution {}

impl Solution {
    pub fn zigzag_level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut res = Vec::<Vec<i32>>::new();
        if let Some(root) = root {
            let mut current_stack = vec![root];
            let mut next_current = Vec::<Rc<RefCell<TreeNode>>>::new();
            let mut level: usize = 0;
            while let Some(node) = current_stack.pop() {
                if current_stack.len() == 0 {
                    
                }
            }
        }
        res
    }
}

#[cfg(test)]
mod zigzag_level_order_test {
    use super::*;

    #[test]
    fn test_1() {
        let n3 = Rc::new(RefCell::new(TreeNode::new(3)));
        let n9 = Rc::new(RefCell::new(TreeNode::new(9)));
        let n20 = Rc::new(RefCell::new(TreeNode::new(20)));
        let n15 = Rc::new(RefCell::new(TreeNode::new(15)));
        let n7 = Rc::new(RefCell::new(TreeNode::new(7)));
        n20.borrow_mut().left = Some(n15);
        n20.borrow_mut().right = Some(n7);
        n3.borrow_mut().left = Some(n9);
        n3.borrow_mut().right = Some(n20);
        assert_eq!(Solution::zigzag_level_order(Some(n3)), vec![vec![3], vec![20,9], vec![15,7]]);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::zigzag_level_order(Some(Rc::new(RefCell::new(TreeNode::new(1))))), vec![vec![1]]);
    }    

    #[test]
    fn test_3() {
        assert_eq!(Solution::zigzag_level_order(None), Vec::<Vec<i32>>::new());
    }    
}